### PR TITLE
Spark: Spark SQL Extensions for create branch

### PR DESCRIPTION
**Pull request by hililiwei**
Co-authored-by: Amogh Jahagirdar <jahamogh@amazon.com>
Co-authored-by: chidayong <247070443@qq.com>

## What is the purpose of the change
Implement the syntax in the following documents:
https://docs.google.com/document/d/1tbATFPrKF3vNlzkgZQdaW8CAJmbjvryfrlg6C2Ci_aA/edit

As suggested by @jackye1995, further split the DDL into smaller PR.

### Create  Branches

```sql
ALTER TABLE tableName
{CREATE | REPLACE} BRANCH branchName [AS OF {VERSION snapshotId}]
[RETAIN interval {DAYS | HOURS | MINUTES}]
[WITH SNAPSHOT RETENTION {[num_snapshots SNAPSHOTS] [interval {DAYS | HOURS | MINUTES}]}]}]

```
e.g.
```
ALTER TABLE prod.db.my_table 
CREATE BRANCH test_branch AS OF VERSION 100 
RETAIN 12 MONTHS
WITH SNAPSHOT RETENTION 5 SNAPSHOTS 7 DAYS
```

cc @jackye1995 @amogh-jahagirdar

### CODE REVIEW

File: spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4
Comment (jackye1995) [inline spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:200]: If we only allow a few values MONTHS, DAYS, HOURS, MINUTES, can we just list them here as keywords instead of using a generic identifier?

File: spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CreateBranchExec.scala
Comment (jackye1995) [inline spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CreateBranchExec.scala:45]: instead of using a hard-coded default, can we only set these config values if they are present? So that the table property default will be used during actual snapshot expiration

File: spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4
Comment (jackye1995) [inline spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:76]: feels more natural for the `RETAIN` clause to be before the `snapshotRetentionClause`? Because clauses like:

```
CREATE BRANCH branch AS OF VERSION 12345 
WITH SNAPSHOT RETENTION 2 SNAPSHOTS 2 DAYS
RETAIN 2 DAYS
```

vs 

```
CREATE BRANCH branch AS OF VERSION 12345 
RETAIN 2 DAYS
WITH SNAPSHOT RETENTION 2 SNAPSHOTS 2 DAYS
```

Making `RETAIN` after `WITH SNAPSHOT RETENTION` makes me feel a bit lost about what that clause is doing. By moving it closer to `CREATE BRANCH`, it is clearer that `RETAIN` is describing expiration of the branch.

What do you think?

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestSnapshotRefSQL.java
Comment (jackye1995) [inline spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestSnapshotRefSQL.java:37]: I think this class name is not updated from the original PR

File: spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4
Comment (amogh-jahagirdar) [reply spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:76]: Agreed, after re-reading the syntax I felt this way too. `RETAIN` should be close to the actual branch clause so that association of what the clauses is responsible for (the retention of the reference itself) is there.

File: spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/CreateBranch.scala
Comment (amogh-jahagirdar) [inline spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/CreateBranch.scala:32]: Nit: spacing in the message 

`"Create branch: ${branch} for table: ${table.quoted}"`

File: spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CreateBranchExec.scala
Comment (amogh-jahagirdar) [inline spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CreateBranchExec.scala:52]: I think this case will be common for all the reference based operations. We may want to see about extracting to a common parent. Not needed at this point, but we may revisit in later DDL implementations.

File: spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4
Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:76]: Sounds reasonable to me.

File: spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4
Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:82]: This skips illegal statements : `WITH SNAPSHOT RETENTION` 
or we can use 
`WITH SNAPSHOT RETENTION ((numSnapshots SNAPSHOTS)? (snapshotRetain snapshotRetainTimeUnit)? | numSnapshots SNAPSHOTS snapshotRetain snapshotRetainTimeUnit )`?
But it's not intuitive

Comment (jackye1995) [reply spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:82]: I see, I will leave it here to see if anyone has better suggestions. I am not an Antlr expert 😝

File: spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4
Comment (jackye1995) [inline spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:212]: missing SECONDS?

File: spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala
Comment (jackye1995) [inline spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala:291]: can use `TimeUnit.MONTHS.toMillis(...)`

File: spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CreateBranchExec.scala
Comment (jackye1995) [reply spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CreateBranchExec.scala:61]: This seems like an existing pattern for all extensions, so I think it is probably fine to leave it like this.

**Comment by jackye1995** [comment]
Ping some people for thoughts around the syntax: @rdblue @RussellSpitzer @nastra

File: spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4
Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:212]: Should we support it? I prefer at least the minute-level.

File: spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala
Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala]: Modified. PTAL. Thanks.

File: spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala
Comment (jackye1995) [inline spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala:291]: hmm, good point, there is not exactly always 30 days in a month. That's probably why it is not supported by `TimeUnit`. So I would say in that case let's remove MONTHS from the time unit in parser and just support DAYS, HOURS, MINUTES and SECONDS

File: spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4
Comment (jackye1995) [reply spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:212]: sure, I am fine with minute level. We can always add more if needed.

File: spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala
Comment (jackye1995) [inline spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala:292]: do we still need to use match statement? The parser will only allow these values, so we should be able to directly parse it to TimeUnit.

File: spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala
Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala]: Yes, it does. Updated.

File: spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala
Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala]: Done.

**Review by jackye1995 — APPROVED** [review]
looks good to me! Waiting for some feedback from other committers

**Review by amogh-jahagirdar — APPROVED** [review]
Took another pass, LGTM! thanks for contributing this

**Review by yyanyy — APPROVED** [review]

**Comment by jackye1995** [comment]
I think this PR is mostly ready to go. I see there is a comment in design doc from @flyrain:

```
"VERSION" is used in Iceberg to indicate any table changes including table property changes, which doesn't create a new snapshot. So versions and snapshots are mismatch. Can we call it "SNAPSHOT" to remove the ambiguity?
```

Requesting one more review from him

**Comment by jackye1995** [comment]
Looks like there are quite a few duplicates for #6637 without merging this one. In that case given there are 2 committer votes, I will first merge this one to unblock that PR, and will request @flyrain 's review on the next PR. Thanks for the work @hililiwei , and thanks for all the reviews @yyanyy and @amogh-jahagirdar !

File: spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4
Comment (rdblue) [inline spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:196]: Why are there so many aliases for number? Are these rules useful?

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java
Comment (rdblue) [inline spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java:42]: Why is this testing with multiple catalogs? This is a table-level operation that shouldn't be affected by the catalog, so it should test just one.

File: spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedDataSourceV2Strategy.scala
Comment (rdblue) [inline spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedDataSourceV2Strategy.scala:66]: Why does this pass the logical plan rather than passing the necessary information? Is it just to avoid a longer line?

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java
Comment (rdblue) [inline spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java:94]: This should assert the state of the branch. I think it would use the current snapshot of main, right?

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java
Comment (rdblue) [inline spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java:65]: This should use a snapshot other than the default to test the clause.

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java
Comment (rdblue) [inline spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java:75]: This needs an assertion about the snapshot referenced by the branch.

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java
Comment (rdblue) [inline spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java:179]: This looks suspicious. Why is this not a mismatched input? Is this even reaching the parser?

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java
Comment (rdblue) [inline spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java:136]: When using variables like this, it really helps to name them with a unit. That makes it easier to validate the uses, like `%d DAYS", ... maxSnapshotAgeDays, ...` and `TimeUnit.DAYS.toMillis(maxSnapshotAgeDays)`.

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java
Comment (rdblue) [inline spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java:185]: `HOURS` and `MINUTES` are never tested, but should be.

File: spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4
Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:196]: @jackye1995  asked the same question.

I originally added snapshotRefRetain and snapshotRetain to make the statement parsing code more readable. Removing it is technically feasible. In the new version, I have removed (including create branch).
ref: https://github.com/apache/iceberg/pull/6637#discussion_r1083582974

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java
Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java:42]: yes, we just test `SparkCatalogConfig.SPARK` catalog.
```
  @Parameterized.Parameters(name = "catalogName = {0}, implementation = {1}, config = {2}")
  public static Object[][] parameters() {
    return new Object[][] {
      {
        SparkCatalogConfig.SPARK.catalogName(),
        SparkCatalogConfig.SPARK.implementation(),
        SparkCatalogConfig.SPARK.properties()
      }
    };
  }
```

File: spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedDataSourceV2Strategy.scala
Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedDataSourceV2Strategy.scala:66]: I passed in the `CreateBranch` instance directly to reduce the argument count.

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java
Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java:185]: I tried using the fourth param to test different units of time.
```
  @Parameterized.Parameters(
      name = "catalogName = {0}, implementation = {1}, config = {2}, timeUnit = {3}")
```

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java
Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java:179]: fixed in the latest version. #6637

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java
Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java:65]: Makes sense. updated in #6637.

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java
Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java:94]: updated in https://github.com/apache/iceberg/pull/6637.

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java
Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java:75]: updated in https://github.com/apache/iceberg/pull/6637.

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java
Comment (hililiwei) [reply spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCreateBranch.java:136]: Changed the unit to a variable. Using a fixed unit suffix is not appropriate, any suggestions?

**Pull request merged**
